### PR TITLE
Cognito auth 1/2 - extracted electron logic

### PR DESCRIPTION
**Pull request by PabloBuchu**
### Pull Request Description

Provides functionality necessary for:
- opening URLs in the system browser (so that we can handle OAuth flows outside of the app)
- handling deep links to the application (so that the OAuth flows can return the user to the app)

### Important Notes

- Modifies `preload.ts` to expose the ability to open the system browser to the sandboxed parts of the app.
- Modifies `election-builder-config.ts` to register a deep link URL protocol scheme with the OS.
- Modifies the client's `index.ts` to register a handler for Electron `open-url` events

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * - The application emits an Electron {@link OPEN_URL_EVENT}.
+ * - The {@link OPEN_URL_EVENT} listener checks if the {@link URL} is a deep link.
+ * - If the {@link URL} is a deep link, the {@link OPEN_URL_EVENT} listener prevents Electron from
+ * handling the event.
+ * - The {@link OPEN_URL_EVENT} listener then emits an {@link ipc.channel.openDeepLink} event.
+ * - The {@link ipc.channel.openDeepLink} listener registered by the dashboard receives the event.
+ * Then it parses the {@link URL} from the event's {@link URL} argument. Then it uses the
+ * {@link URL} to redirect the user to the dashboard, to the page specified in the {@link URL}'s
+ * `pathname`. */
+import * as electron from 'electron'
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts:76]: Add 1 empty line before imports pls.

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    electron.ipcMain.on(ipc.channel.openUrlInSystemBrowser, (_event, url) => opener(url))
+}
+
+/** Registers a listener that fires a callback for `open-url` events, when the URL is a deep link.
+ *
+ * This listener is used to open a page in *this* application window, when the user is
+ * redirected to a URL with a protocol supported by this application.
+ *
+ * All URLs that aren't deep links (i.e., URLs that don't use the {@link shared.DEEP_LINK_SCHEME}
+ * protocol) will be ignored by this handler. */
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts]: I was commenting it in my previous PR review - ignoring non deep links can cause REALLY long debugging process if something goes wrong. We should never ignore bad things. They should always be reported as warnings or errors.

File: app/ide-desktop/lib/client/src/index.ts [hunk 3]
```diff
@@ -50,7 +51,7 @@ class App {
             this.setChromeOptions(chromeOptions)
             security.enableAll()
             electron.app.on('before-quit', () => (this.isQuitting = true))
-            electron.app.whenReady().then(() => this.main(windowSize))
+            electron.app.on('ready', () => this.main(windowSize))
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/index.ts]: Why this change? After this change this works way worse than before this change. Your version does not work if the `'ready'` event was already emitted. The previous version was working fine then. Anyway, I'd love to understand why such a change was made, as such changes can cause potentially big problems and we need to be very careful with them.

File: app/ide-desktop/lib/client/src/index.ts [hunk 4]
```diff
@@ -102,6 +103,7 @@ class App {
                 await this.startContentServerIfEnabled()
                 await this.createWindowIfEnabled(windowSize)
                 this.initIpc()
+                authentication.initAuthenticationModule(() => this.window)
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/index.ts]: 1. We are repeating the "authentication" name here. It should be probably removed from the function name, as the function was defined in the "authentication" module.
2. Why the window is passed as lambda and not as a variable?

File: app/ide-desktop/lib/client/src/preload.ts [hunk 5]
```diff
@@ -6,6 +6,18 @@
...
 import * as ipc from 'ipc'
 
+
+
+// =================
+// === Constants ===
+// =================
+
+/** Name of the object containing proxied authentication functions that are used by the
+ * dashboard. */
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/preload.ts]: I do not understand this documentation. What is "the object" and what are "proxified functions"? Where is the object stored?

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * This channel listens for {@link ipc.channel.openUrlInSystemBrowser} events. When this kind of
+ * event is fired, this listener will assume that the first and only argument of the event is a URL.
+ * This listener will then attempt to open the URL in a cross-platform way. The intent is to open
+ * the URL in the system browser.
+ *
+ * This functionality is necessary because we don't want to run the OAuth flow in the app. Users
+ * don't trust Electron apps to handle their credentials. */
+const initIpc = () => {
+    electron.ipcMain.on(ipc.channel.openUrlInSystemBrowser, (_event, url) => opener(url))
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts:117]: I was commenting it in my previous PR review - this is the only usage of `opener`. Why are we importing whole library just for opening a url? Can we just use built-in methods for it?

File: app/ide-desktop/lib/client/shared.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+ * - the name of the product, and
+ * - custom URL protocol scheme definitions.
+ *
+ * This metadata is used in both the code building the client resources and the packaged code
+ * itself. */
+
+/** Name of the product. */
+export const PRODUCT_NAME = 'Enso'
+
+/** URL protocol scheme for deep links to authentication flow pages. */
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/shared.ts]: Please add here an example, showing that this will be used as `enso://...`

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * The main Electron process can launch the system web browser. The dashboard and IDE are sandboxed,
+ * so they can not launch the system web browser. By registering Inter-Process Communication (IPC)
+ * listeners in the Electron app, we can bridge this gap, and allow the dashboad + IDE to emit
+ * events that signal to the main Electron process to open URLs in the system web browser.
+ *
+ * ## Redirect To System Web Browser
+ *
+ * The user must use the system browser to complete sensitive flows such as signup and signin. These
+ * flows should not be done in the app as the user cannot be expected to trust the app with their
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts:38]: this doc is beautiful and very clear. Thanks for writing it

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+ * redirected to a URL with a protocol supported by this application.
+ *
+ * All URLs that aren't deep links (i.e., URLs that don't use the {@link shared.DEEP_LINK_SCHEME}
+ * protocol) will be ignored by this handler. */
+const initOpenUrlListener = (
+    window: () => electron.BrowserWindow | null,
+) => {
+    electron.app.on(OPEN_URL_EVENT, (event, url) => {
+        const parsedUrl = new URL(url)
+
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts]: why these newlines?

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        if (parsedUrl.protocol !== `${shared.DEEP_LINK_SCHEME}:`) {
+            return
+        }
+
+        /** Don't open the deep link URL in the window, we want the system browser to handle it. */
+        event.preventDefault()
+
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts:143]: why these newlines?

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+const initOpenUrlListener = (
+    window: () => electron.BrowserWindow | null,
+) => {
+    electron.app.on(OPEN_URL_EVENT, (event, url) => {
+        const parsedUrl = new URL(url)
+
+        if (parsedUrl.protocol !== `${shared.DEEP_LINK_SCHEME}:`) {
+            return
+        }
+
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts:139]: why these newlines?

File: app/ide-desktop/lib/client/src/authentication.ts [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        /** Don't open the deep link URL in the window, we want the system browser to handle it. */
+        event.preventDefault()
+
+        window()?.webContents.send(ipc.channel.openDeepLink, url)
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts]: What if window does not exist? When can it happen? Should we report errors/warnings then?

File: app/ide-desktop/lib/client/src/ipc.ts [hunk 13]
```diff
@@ -12,4 +14,10 @@ export const channel = {
     profilesLoaded: 'profiles-loaded',
     saveProfile: 'save-profile',
     quit: 'quit-ide',
+    /** Channel for requesting that a URL be opened by the system browser. */
+    openUrlInSystemBrowser: 'open-url-in-system-browser',
+    /** Channel for setting a callback that handles deep links to this application. */
+    setDeepLinkHandler: 'set-deep-link-handler',
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/ipc.ts:18]: I don't see this variable used anywhere in the code.

File: app/ide-desktop/lib/client/src/preload.ts [hunk 14]
```diff
@@ -50,3 +62,28 @@ contextBridge.exposeInMainWorld('enso_console', {
...
+     * OAuth URLs must be opened this way because the dashboard application is sandboxed and thus
+     * not privileged to do so unless we explicitly expose this functionality. */
+    openUrlInSystemBrowser: (url: string) => ipcRenderer.send(ipc.channel.openUrlInSystemBrowser, url),
+    /** Set the callback that will be called when a deep link to the application is opened.
+     *
+     * The callback is intended to handle links like
+     * `enso://authentication/register?code=...&state=...` from external sources like the user's
+     * system browser or email client. Handling the links involves resuming whatever flow was in
+     * progress when the link was opened (e.g., an OAuth registration flow). */
+    setDeepLinkHandler: (callback: (url: string) => void) =>
```
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/preload.ts:89]: this is not used anyhweher in the code. I understand that it will be used by the subsequent PR, right?

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 2 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts]: *This* handler is only interested in deep links. So by ignoring the non deep links, they end up having the default behaviour, which is to let Electron itself handle them. Is that not what we want? We're not disabling any functionality, only adding a special case for deep links.

File: app/ide-desktop/lib/client/src/index.ts (hunk 4 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/index.ts]: It's passed as a lambda and not a module because the window is uninitialized at this point. If we passed it as a variable it'd remain unset.

We only need to use it from the dashboard/IDE, so it's only needed after initialization, so it's fine if it's unset when the function is called.

File: app/ide-desktop/lib/client/src/preload.ts (hunk 5 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/preload.ts]: This is partially explained later in the file, when this variable is used:
```typescript
/** Exposes an object on the main window; object is used by the dashboard to:
 * - open OAuth flows in the system browser, and
 * - handle deep links from the system browser or email client to the dashboard. */
contextBridge.exposeInMainWorld(AUTHENTICATION_API_KEY, {
```

Should I duplicate the documentation or move it to this variable?

As for proxied functions, I'll add documentation explaining that.

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 6 above)
Comment (PabloBuchu) [reply app/ide-desktop/lib/client/src/authentication.ts:117]: So what `opener` does is it opens a browser with requested url in it. The library is a slim wrapper that detects on which platform is being used and calls right command:
- Windows: `cmd.exe`
- Mac: open
- Linux: xdg-open

In fact it's just couple lines of code: https://github.com/domenic/opener/blob/master/lib/opener.js

I also don't like installing a lib for single purpose only but in this case this is a useful snippet of code I think.

If you really insist I would vote for creating FIXME with a chore ticket not to block the whole PR

Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts:117]: I responded to that comment in the previous PR review. This library is ~60 LOC with 0 dependencies and it's used to make the opening of URLs work cross-platform. Yes, we could use built-in methods, but it'd be easier to vendor the dependency into the repo since we'd end up writing the same code. Is that OK?

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 2 above)
Comment (PabloBuchu) [reply app/ide-desktop/lib/client/src/authentication.ts]: Do you know any nice way of providing here a logger instance to enable logging or should I use just console.log?

File: app/ide-desktop/lib/client/src/index.ts (hunk 4 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/index.ts]: I suppose it's worth documenting this since it's non-obvious.

File: app/ide-desktop/lib/client/src/index.ts (hunk 3 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/index.ts]: IIRC without this change `open-url` events didn't fire when deep links were clicked. I'm still not 100% sure _why_ that happened. I'll investigate the issue and try to resolve it without this change, if possible. I'll document why this change was made, if it isn't.

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 9 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts]: Seemed nicer to me 🤷, will change

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 10 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts:143]: To make it clear that we perform the side effect of disabling the default behaviour at this point, before proceeding with the important part: handling the link. Can change this.

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 11 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts:139]: Again, to separate the `event.preventDefault()`, which is more of a side-effect than the intended logic.

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 12 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts]: I'm not sure if it's possible for the window to not exist at this point. Will have to read the Electron docs to see when the `open-url` event can fire. I think this may be related to the `.on('ready', ...)` vs `whenReady()` change.

File: app/ide-desktop/lib/client/src/ipc.ts (hunk 13 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/ipc.ts:18]: This will be used by the subsequent PR. The dashboard is responsible for setting a handler, since it's the module that knows where/how to route the authentication pages.

File: app/ide-desktop/lib/client/src/preload.ts (hunk 14 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/preload.ts:89]: This will be used by the subsequent PR. The dashboard is responsible for setting a handler, since it's the module that knows where/how to route the authentication pages.

File: app/ide-desktop/lib/client/src/preload.ts (hunk 5 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/preload.ts]: The term "proxied functions" comes from the Electron docs, FYI: <https://www.electronjs.org/docs/latest/api/context-bridge#api-functions>

> Function values that you bind through the contextBridge are proxied through Electron to ensure that contexts remain isolated. This results in some key limitations that we've outlined below.

File: app/ide-desktop/lib/client/src/preload.ts (hunk 14 above)
Comment (PabloBuchu) [reply app/ide-desktop/lib/client/src/preload.ts:89]: Yes thats correct

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 12 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/authentication.ts]: Updated the docs & lambda to account for this scenario.

File: app/ide-desktop/lib/client/src/index.ts (hunk 3 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/index.ts]: Yeah so without this change the application freezes on startup. The spinner gets and stays stuck. I'm unable to diagnose why since I'm not able to get `showLogs()` to work once this happens. Any advice on how to investigate the cause of such a crash? @wdanilo

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 2 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/authentication.ts]: @indiv0 we should not allow Electron to handle any redirections, as it could potentially be dangerous for users. So, we should allow redirection of deep links and prevent any other redirection while logging an error. 

@PabloBuchu `client/src/index.ts` has logger access: `const logger = content.logger` - so you should be able to access it as well. Does it solve the issue?

File: app/ide-desktop/lib/client/src/index.ts (hunk 3 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/index.ts]: Super strange. Let's do this:

1. Add this code:

```
electron.app.whenReady().then(() => console.log("whenReady!"))
electron.app.on('ready', () => console.log('ready'))
```

2. Leave the version with `whenReady` and try replicating the app stuck on loader screen.
3. Tell me what you see in terminal.
4. Until its resolved, lets change it to `on('ready', ...)` + lets create issue for it on our issue tracker and link it in the codebase.
5. When we will have that confirmed / diagnosed, we will report that in electron's GitHub.

File: app/ide-desktop/lib/client/src/preload.ts (hunk 5 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/preload.ts]: 1. If this is referring an anonymous object that is defined below, let's make the object non-anonymous (just assign it to var before passing it to `contextBridge.exposeInMainWorld` and then refer to it by name from the docs.
2. It would be good to put the object above this function, so the file can be read from top-to bottom without the need to jump to understand things.

Do you like this solution? :)

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 6 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/authentication.ts:117]: Oh, so this is actually doing some interesting logic there, like it's platform-specific. Ok, in such a case my comment is invalid. Thanks for pointing it out!

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 11 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/authentication.ts:139]: In such cases in our Rust codebase, we are just refactoring the body of the function to a few, well named sub-functions, so reading them line-by-line explains everything there. We can leave it this way, but this is an exception and we are not really using such newlines in other parts of the codebase.

File: app/ide-desktop/lib/client/src/authentication.ts (hunk 12 above)
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/authentication.ts]: We should always log error / warning so if it happens some day, we will have easier time debugging it.

**Review by wdanilo — APPROVED** [review]

**Comment by wdanilo** [comment]
Approved! But please apply the things mentioned in the latests comments :)

File: app/ide-desktop/lib/client/src/preload.ts (hunk 5 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/preload.ts]: I see what you mean. That's a clever solution. I'll add it.

File: app/ide-desktop/lib/client/src/index.ts (hunk 3 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/src/index.ts]: Sounds good. Opened as issue #5851 and added a comment to the codebase.
